Tasks/Gradle/sonarqube-common/sonarqube-rest.ts

```diff
@@ -0,0 +1,193 @@
+                if (!responseJson.projectStatus) {
+                    throw responseJson;
+                }
+
+                return responseJson;
+            });
+    } catch (error) {
+        tl.debug('Could not fetch quality gate details on analysis ID' + analysisId);
+        throw error;
+    }
+}
+
+function getTaskDetails(taskReport:TaskReport):Q.Promise<Object> {
+    if (!taskReport) {
+        throw new Error(tl.loc('sqAnalysis_TaskReportInvalid'));
+    }
+
+    try {
+        return callSonarQubeRestEndpoint('/api/ce/task?id=' + taskReport.ceTaskId)
+            .then((response:RestResponse) => {
+                var responseJson:any = response.payloadToJson();
+                if (!responseJson.task) {
+                    throw responseJson;
+                }
+
+                return responseJson;
+            });
+    } catch (error) {
+        tl.debug('Could not fetch task details on ID' + taskReport.ceTaskId);
+        throw error;
+    }
+}
+
+// Invokes a REST endpoint at the SonarQube server.
+function callSonarQubeRestEndpoint(path:string):Q.Promise<RestResponse> {
+    var defer = Q.defer<RestResponse>();
+
+    var options:any = createSonarQubeHttpRequestOptions(path);
+
+    // Dynamic switching - we cannot use https.request() for http:// calls or vice versa
+    var protocolToUse;
+    switch (options.protocol) {
+        case 'http':
+            protocolToUse = http;
+            break;
+        case 'https':
+            protocolToUse = https;
+            break;
+        default:
+            protocolToUse = http;
+            break;
+    }
+
+    var responseBody:string = '';
+    var request = protocolToUse.request(options, (response:IncomingMessage) => {
+
+        response.on('data', function (body) {
+            responseBody += body;
+        });
+
+        response.on('end', function () {
+            var serverResponseString:string = response.statusCode + " " + http.STATUS_CODES[response.statusCode];
+            tl.debug('Got response: ' + serverResponseString + " from " + path);
+            tl.debug(responseBody);
+            var result:RestResponse = new RestResponse(response.statusCode, responseBody);
+            if (!result.wasSuccess()) {
+                defer.reject('Server responded with ' + serverResponseString);
+            } else {
+                defer.resolve(result);
+            }
+        });
+    });
+
+    request.on('error', (error) => {
+        tl.debug('Failed to call ' + path);
+        defer.reject(error);
+    });
+
+    tl.debug('Sending request to: ' + path);
+    request.end();
+    return defer.promise;
+}
+
+// Constructs the options object used by the https.request() method. Takes the host-relative path i.e. '/' as an argument.
+function createSonarQubeHttpRequestOptions(path?:string):Object {
+    const endpoint:SonarQubeEndpoint = sqCommon.getSonarQubeEndpoint();
+
+    var hostUrl:url.Url = url.parse(endpoint.Url);
```

Comment:
_url.parse()_ can tolerate strange characters - it simply escapes them as per the URI standard. I have not observed it throw an exception, and I don't see any (realistic and legitimate) failure scenario.